File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/memory/search/HybridSearch.kt
/** Hybrid search — combines FTS5 (BM25), vector similarity, graph context, and tier boost. */
package com.codeintel.memory.search

import com.codeintel.memory.embedding.EmbeddingService
import com.codeintel.memory.graph.KnowledgeGraph
import com.codeintel.memory.models.KnowledgeSearchResult
import com.codeintel.memory.repository.KnowledgeSearchRepository
import com.codeintel.memory.repository.VectorRepository

/** Search parameters for hybrid search. */
data class SearchParams(
    val query: String,
    val limit: Int = 10,
    val tier: String? = null,
    val type: String? = null,
    val role: String? = null,
    val useVector: Boolean = true,
    val useGraph: Boolean = true,
    val bm25Weight: Double = 0.4,
    val vectorWeight: Double = 0.4,
    val graphWeight: Double = 0.2
)

/** Enhanced search parameters with agent scope and token budget. */
data class EnhancedSearchParams(
    val query: String,
    val limit: Int = 10,
    val tier: String? = null,
    val type: String? = null,
    val role: String? = null,
    val agentScope: String? = null,
    val maxTokens: Int = 2000
)

/** Enhanced search response with pinned context, budget info, and expiry actions. */
data class EnhancedSearchResponse(
    val pinnedContext: String,
    val results: List<KnowledgeSearchResult>,
    val tokensUsed: Int,
    val tokensBudget: Int,
    val resultsTruncated: Boolean,
    val expiryActions: List<ExpiryAction>
)

class HybridSearch(
    private val ftsRepo: KnowledgeSearchRepository,
    private val vectorRepo: VectorRepository,
    private val embeddingService: EmbeddingService?,
    private val graph: KnowledgeGraph?
) {
    // V2 injectable dependencies (KSA-110 F4)
    private var scopeFilter: AgentScopeFilter? = null
    private var tokenBudget: TokenBudget? = null
    private var workingExpiry: WorkingTierExpiry? = null
    private var pinnedContextProvider: (() -> String)? = null

    /** Inject AgentScopeFilter for tag-based isolation. */
    fun setScopeFilter(filter: AgentScopeFilter) { this.scopeFilter = filter }

    /** Inject TokenBudget for result limiting. */
    fun setTokenBudget(budget: TokenBudget) { this.tokenBudget = budget }

    /** Inject WorkingTierExpiry for lazy auto-expiry. */
    fun setWorkingExpiry(expiry: WorkingTierExpiry) { this.workingExpiry = expiry }

    /** Inject pinned context provider (CoreMemoryManager.getContext). */
    fun setPinnedContextProvider(provider: () -> String) { this.pinnedContextProvider = provider }

    /** Enhanced search: expiry → pins → search → scope → budget. */
    fun enhancedSearch(params: EnhancedSearchParams): EnhancedSearchResponse {
        // 1. Lazy auto-expiry
        val expiryActions = workingExpiry?.processStale() ?: emptyList()

        // 2. Load pinned context
        val pinnedContext = pinnedContextProvider?.invoke() ?: ""

        // 3. Execute hybrid search
        val searchParams = SearchParams(
            query = params.query, limit = params.limit,
            tier = params.tier, type = params.type, role = params.role
        )
        var results = search(searchParams)

        // 4. Apply agent scope filter
        if (params.agentScope != null && scopeFilter != null) {
            results = scopeFilter!!.filter(results, params.agentScope)
        }

        // 5. Apply token budget
        val maxTokens = params.maxTokens
        var tokensUsed = 0
        var truncated = false
        if (tokenBudget != null) {
            val budgetResult = tokenBudget!!.apply(results, maxTokens)
            results = budgetResult.results
            tokensUsed = budgetResult.tokensUsed
            truncated = budgetResult.truncated
        }

        return EnhancedSearchResponse(pinnedContext, results, tokensUsed, maxTokens, truncated, expiryActions)
    }

    /** Execute hybrid search with RRF fusion + tier boost + role filter. */
    fun search(params: SearchParams): List<KnowledgeSearchResult> {
        val ftsResults = searchFts(params)
        val vectorResults = if (params.useVector) searchVector(params) else emptyMap()
        val graphBoost = if (params.useGraph) computeGraphBoost(ftsResults, vectorResults) else emptyMap()
        return fuseResults(ftsResults, vectorResults, graphBoost, params)
    }

    private fun searchFts(params: SearchParams): Map<Long, KnowledgeSearchResult> {
        val results = if (params.tier != null) {
            ftsRepo.searchInTier(params.query, params.tier, params.limit * 2)
        } else {
            ftsRepo.search(params.query, params.limit * 2)
        }
        return results.associateBy { it.entry.id }
    }

    private fun searchVector(params: SearchParams): Map<Long, Double> {
        val queryVec = embeddingService?.embed(params.query) ?: return emptyMap()
        val allVectors = vectorRepo.findAll()
        return allVectors
            .map { it.entryId to EmbeddingService.cosine(queryVec, EmbeddingService.bytesToFloatArray(it.vector)) }
            .sortedByDescending { it.second }
            .take(params.limit * 2)
            .toMap()
    }

    private fun computeGraphBoost(
        ftsResults: Map<Long, KnowledgeSearchResult>,
        vectorResults: Map<Long, Double>
    ): Map<Long, Double> {
        val g = graph ?: return emptyMap()
        val seedIds = (ftsResults.keys + vectorResults.keys).take(5)
        val boosted = mutableMapOf<Long, Double>()
        for (seedId in seedIds) {
            val neighbors = g.getConnected(seedId)
            for (n in neighbors) { boosted[n] = (boosted[n] ?: 0.0) + 1.0 }
        }
        val maxBoost = boosted.values.maxOrNull() ?: 1.0
        return boosted.mapValues { it.value / maxBoost }
    }

    private fun fuseResults(
        ftsResults: Map<Long, KnowledgeSearchResult>,
        vectorScores: Map<Long, Double>,
        graphBoost: Map<Long, Double>,
        params: SearchParams
    ): List<KnowledgeSearchResult> {
        val allIds = ftsResults.keys + vectorScores.keys + graphBoost.keys
        val roleTypes = RoleFilter.typesForRole(params.role)
        val scored = allIds.mapNotNull { id ->
            val entry = ftsResults[id]?.entry ?: return@mapNotNull null
            if (roleTypes != null && entry.type !in roleTypes) return@mapNotNull null
            val ftsScore = rrfScore(ftsResults.keys.indexOf(id)) * params.bm25Weight
            val vecScore = (vectorScores[id] ?: 0.0) * params.vectorWeight
            val gScore = (graphBoost[id] ?: 0.0) * params.graphWeight
            val tierBoost = TierBoost.factor(entry.tier)
            val total = (ftsScore + vecScore + gScore) * tierBoost
            Triple(id, total, entry)
        }
        return scored
            .sortedByDescending { it.second }
            .take(params.limit)
            .map { KnowledgeSearchResult(it.third, it.second, "hybrid") }
    }

    /** Reciprocal Rank Fusion score. */
    private fun rrfScore(rank: Int, k: Int = 60): Double {
        if (rank < 0) return 0.0
        return 1.0 / (k + rank + 1)
    }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/memory/hybrid_search.py
"""HybridSearch — combines FTS5 (BM25) and graph context with RRF fusion.

Enhanced with: agent scope filter, token budget, working tier expiry (KSA-110 F4).
"""

from dataclasses import dataclass
from typing import Any

from .search_repo import KnowledgeSearchRepository
from .knowledge_graph import KnowledgeGraph
from .role_filter import types_for_role
from .tier_boost import factor as tier_boost_factor


@dataclass
class EnhancedSearchResponse:
    """Enhanced search response with pinned context, budget info, and expiry actions."""

    pinned_context: str
    results: list[dict[str, Any]]
    tokens_used: int
    tokens_budget: int
    results_truncated: bool
    expiry_actions: list


class HybridSearch:
    """Hybrid search with RRF fusion (BM25 + graph boost + role filter + tier boost)."""

    def __init__(self, fts_repo: KnowledgeSearchRepository,
                 graph: KnowledgeGraph) -> None:
        self._fts = fts_repo
        self._graph = graph
        # V2 injectable dependencies (KSA-110 F4)
        self._scope_filter = None
        self._token_budget = None
        self._working_expiry = None
        self._pinned_context_provider = None

    def set_scope_filter(self, filter_) -> None:
        """Inject AgentScopeFilter for tag-based isolation."""
        self._scope_filter = filter_

    def set_token_budget(self, budget) -> None:
        """Inject TokenBudget for result limiting."""
        self._token_budget = budget

    def set_working_expiry(self, expiry) -> None:
        """Inject WorkingTierExpiry for lazy auto-expiry."""
        self._working_expiry = expiry

    def set_pinned_context_provider(self, provider) -> None:
        """Inject pinned context provider (callable returning str)."""
        self._pinned_context_provider = provider

    def enhanced_search(self, query: str, limit: int = 10,
                        tier: str | None = None, role: str | None = None,
                        agent_scope: str | None = None,
                        max_tokens: int = 2000) -> EnhancedSearchResponse:
        """Enhanced search: expiry → pins → search → scope → budget."""
        # 1. Lazy auto-expiry
        expiry_actions = []
        if self._working_expiry:
            expiry_actions = self._working_expiry.process_stale()

        # 2. Load pinned context
        pinned_context = ""
        if self._pinned_context_provider:
            pinned_context = self._pinned_context_provider()

        # 3. Execute hybrid search
        results = self.search(query, limit, tier, role)

        # 4. Apply agent scope filter
        if agent_scope and self._scope_filter:
            results = self._scope_filter.filter(results, agent_scope)

        # 5. Apply token budget
        tokens_used = 0
        truncated = False
        if self._token_budget:
            budget_result = self._token_budget.apply(results, max_tokens)
            results = budget_result.results
            tokens_used = budget_result.tokens_used
            truncated = budget_result.truncated

        return EnhancedSearchResponse(
            pinned_context=pinned_context,
            results=results,
            tokens_used=tokens_used,
            tokens_budget=max_tokens,
            results_truncated=truncated,
            expiry_actions=expiry_actions,
        )

    def search(self, query: str, limit: int = 10,
               tier: str | None = None, role: str | None = None) -> list[dict[str, Any]]:
        """Execute hybrid search."""
        fts_results = self._search_fts(query, tier, limit * 2)
        graph_boost = self._compute_graph_boost(fts_results)
        return self._fuse(fts_results, graph_boost, limit, role)

    def _search_fts(self, query: str, tier: str | None,
                    limit: int) -> list[dict[str, Any]]:
        if tier:
            return self._fts.search_in_tier(query, tier, limit)
        return self._fts.search(query, limit)

    def _compute_graph_boost(self, fts_results: list[dict[str, Any]]) -> dict[int, float]:
        seed_ids = [r["entry"]["id"] for r in fts_results[:5]]
        boosted: dict[int, float] = {}
        for seed_id in seed_ids:
            for n in self._graph.get_connected(seed_id):
                boosted[n] = boosted.get(n, 0.0) + 1.0
        max_boost = max(boosted.values(), default=1.0)
        return {k: v / max_boost for k, v in boosted.items()}

    def _fuse(self, fts_results: list[dict[str, Any]],
              graph_boost: dict[int, float], limit: int,
              role: str | None = None) -> list[dict[str, Any]]:
        bm25_weight = 0.6
        graph_weight = 0.4
        role_types = types_for_role(role)
        fts_map = {r["entry"]["id"]: (r, i) for i, r in enumerate(fts_results)}
        all_ids = set(fts_map.keys()) | set(graph_boost.keys())

        scored: list[tuple[float, dict[str, Any] | None]] = []
        for entry_id in all_ids:
            fts_entry = fts_map.get(entry_id)
            if fts_entry is None:
                continue
            result, rank = fts_entry
            entry_type = result.get("entry", {}).get("type", "")
            if role_types and entry_type not in role_types:
                continue
            fts_score = self._rrf_score(rank) * bm25_weight
            g_score = graph_boost.get(entry_id, 0) * graph_weight
            entry_tier = result.get("entry", {}).get("tier")
            boost = tier_boost_factor(entry_tier)
            total = (fts_score + g_score) * boost
            scored.append((total, result))

        scored.sort(key=lambda x: x[0], reverse=True)
        output: list[dict[str, Any]] = []
        for score, result in scored[:limit]:
            if result is not None:
                output.append({**result, "score": score, "match_type": "hybrid"})
        return output

    def _rrf_score(self, rank: int, k: int = 60) -> float:
        """Reciprocal Rank Fusion score."""
        return 1.0 / (k + rank + 1)
